### PR TITLE
Add dashboard URL for ensembling jobs.

### PR DESCRIPTION
**Pull request by ashwinath**
This PR adds the monitoring dashboard URL for the ensembling job. All ensembling jobs have the same URL so there isn't a need to persist into the database, so it is populated at runtime and sent back during the API call.

Affected APIs:
- create ensembling job
- list ensembling jobs
- get ensembling job

### CODE REVIEW

File: api/api/specs/jobs.yaml
Comment (krithika369) [inline api/api/specs/jobs.yaml:163]: This `EnsemblingJob` is also used as the request body for the create job request. But we don't expect the dashboard URL to be input (the same problem also exists with the `id` field and perhaps others in the create request).

I think we can define an explicit type to handle the create request. Eg:

```
EnsemblingJob:
      allOf:
        - $ref: '#/components/schemas/EnsemblingJobCreateRequestBody'
        - type: object
          properties:
            id:
              $ref: "#/components/schemas/Id"
            dashboard_url:
              type: string
```

wdyt ?

File: api/turing/service/ensembling_job_service.go
Comment (krithika369) [inline api/turing/service/ensembling_job_service.go:191]: I would've expected that this URL is formatted with the project / namespace / job info for the specific job. Do you have an example of how and where this URL is expected to be used?

For the real time routers, we set the templated URL in the UI config which is populated according to the router being viewed. We've had some discussions in the past, for the real-time router dashboard URL to be moved to the API layer as well (see task id 634 in our backlog), so it may also be available to the SDK. So, I agree with adding the ensembling job's dashboard URL to the API. My concern is just with returning the same value for every job - that feels incomplete to me.

File: api/api/specs/jobs.yaml
Comment (ashwinath) [reply api/api/specs/jobs.yaml:163]: Seems like this problem was solved using the `readOnly: true`, will do that instead. example: https://github.com/ashwinath/turing/blob/f2ea45af2f49114aff304be2585ba6095fb8d3bd/api/api/specs/routers.yaml#L414

Comment (krithika369) [reply api/api/specs/jobs.yaml:163]: Nice. TIL. :)

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:191]: This makes sense, I think this should be implemented, feels rather weird when I first did it as well. Let me add this.

Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go]: @krithika369 I have made the changes to allow templating.

File: api/db-migrations/000009_ensembling_job.up.sql
Comment (krithika369) [inline api/db-migrations/000009_ensembling_job.up.sql:25]: Do we need to persist this in the DB now? In a scenario where the dashboard URL is modified (although rare), we will have to update the data in the DB. For this reason, I think it is better to format the template string dynamically.

File: api/turing/config/example.yaml
Comment (krithika369) [inline api/turing/config/example.yaml:17]: Can we add more details here on what templating vars are supported exactly (is it only these 2?) and what they will be replaced with? Like `{{.Project}}` and `{{.Job}}` for project name and job name, respectively.

File: api/turing/service/ensembling_job_service.go
Comment (krithika369) [inline api/turing/service/ensembling_job_service.go:25]: Ohh why this limit? I'm afraid having this constraint is not very generalisable. Is it a limitation from the internal dashboard?

File: api/db-migrations/000009_ensembling_job.up.sql
Comment (ashwinath) [reply api/db-migrations/000009_ensembling_job.up.sql:25]: I thought about it, it's quite troublesome because a few endpoints are affected:
1. create ensembling job
2. list ensembling job
3. get ensembling job

You might get a runtime performance hit (maybe negligible) when you call the list ensembling job method, since you have to construct the url for every job. I picked the persistence method because less code was required.

Comment (ashwinath) [reply api/db-migrations/000009_ensembling_job.up.sql:25]: But if you feel strongly about this we can always change it to a runtime thing. I think the data update might be relevant but ensembling jobs are a temporary thing, the dashboard wouldn't be very useful later on, i doubt it will be worth it to even update the values.

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:25]: It's a limit because of how spark truncates the pod name, I don't remember the exact value because I used trial and error, the value could be more but 20 seems safe. Are you suggesting hardcoding it?

File: api/turing/config/example.yaml
Comment (ashwinath) [reply api/turing/config/example.yaml]: Yep ok

File: api/db-migrations/000009_ensembling_job.up.sql
Comment (ashwinath) [reply api/db-migrations/000009_ensembling_job.up.sql:25]: actually after thinking about it, I'll change it to generate during runtime. so that its always independent of config.

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:25]: I think there are certain implications that were overlooked here as discussed with @krithika369 offline.

This setting is reversed engineering from how spark generates the pod names. Generally speaking, there are a few components at play here and I will try my best to explain what is generated in terms of order.

1. Ensembling job name provided by the user is appended with the unix timestamp appended to the back. This is currently limited to 50 characters excluding the timestamp (60 included).
2. SparkApplication (CRD) is created with the name equivalent to ensembling job name which has a maximum of 60 characters.
3. Spark Driver Pod is created with the Pod name of a format roughly equivalent to this: `<Ensembling Job Name truncated to some length>-<generated hash>-driver`
4. Spark Executor Pod is created with the Pod name of a format roughly equivalent to this: `<Ensembling Job Name truncated to some length>-<generated hash>-exec-<replica number>`

So the dashboard works by using the metric provided by kube-state-metrics: `kube_pod_container_status_ready` and other relevant metric names by filtering it based on `pod=~"$job.*-driver|$job.*-exec-.*"`. According to the [kube-state-metrics documentation](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md) there seems to be no filter based on a pod annotation/label so this becomes extremely awkward and the reason why I've decided to do a string operation to extract the value out of it.

Now the problem is how are we expecting our users to name their ensembling jobs? @krithika369 probably suspects that they are going to create a name based on a fixed portion appended by some generated string; i.e. in the format `<fixed name>-<generated>` which might be a problem if we truncate to such short character sets of length 20.

My proposal is to limit the number of characters in the CreateEnsemblingJob process to 20 characters. This probably makes things a little tough. If we append 10 characters for the timestamp, users are basically left with 10 characters for their ensembling job name. A few approaches I can think of:

1. Recommend them to use a name that is short and tell them there will be a timestamp generated at the end of it, limiting them to 10 characters only. 
2. Leave the responsibility of letting them generate a name themselves.

In both circumstances, we would have to enforce a unique index on the name + mlp project or there might be collisions in naming.

Honestly speaking, I'm not sure what should be the right way to approach this problem. Maybe @romanwozniak you could weigh in on this.

Comment (krithika369) [reply api/turing/service/ensembling_job_service.go:25]: > My proposal is to limit the number of characters in the CreateEnsemblingJob process to 20 characters

After thinking about this a bit and discussing with @ashwinath offline, limiting the Ensembler name length sounds like a reasonable approach to me (for comparison, we set a 25 chart limit for [router names](https://github.com/gojek/turing/blob/main/ui/src/router/components/form/validation/schema.js#L214) which is probably the reason why we don't have to worry about any truncation of the k8s resource name later). We could put it the same constraint for ensembler names. And similar to the router version number, we can introduce the concept of a `run-id` in the jobs table that is unique per `(MLP Project ID, Ensembler ID)` so that the final name of the job (ensembler name + run ID, instead of name + timestamp) is still < 30 characters and allows enough room for the other suffixes appended by the Spark op on its resources.

> In both circumstances, we would have to enforce a unique index on the name + mlp project or there might be collisions in naming.

Makes sense to me. We do a unique constraint on (MLP Project ID, Router Name). I think the same logic would work for Ensemblers - (MLP Project ID, Ensembler Name).

Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:25]: This also affects the ensembler name as we create the ensembling job name based on the ensembler name.

File: api/turing/service/ensembling_job_service.go
Comment (romanwozniak) [reply api/turing/service/ensembling_job_service.go]: Can we also use `DashboardURL` -> `MonitoringURL` to align it with Merlin ([here](https://github.com/gojek/merlin/blob/805f33fc67ab17f23a85b1b89283b2f12d100fbf/api/models/version_endpoint.go#L38))?

File: api/turing/service/ensembling_job_service.go
Comment (romanwozniak) [reply api/turing/service/ensembling_job_service.go:25]: RunID approach, proposed by @krithika369 makes sense to me. So then we can limit ensembler's name length to 25 char (to make it consistent with routers) and append sequential run id. But propose to implement it in a separate PR

File: api/turing/api/ensembling_job_api.go
Comment (ashwinath) [reply api/turing/api/ensembling_job_api.go:53]: Hmm this makes things quite awkward. From what I understand from the docs, the implementation would look something like this:

```go
func (j *EnsemblingJob) AfterFind() error {
  proj, err := MLPServiceSingleton.GetProject(j.ProjectID) // <------- this is where it gets awkward, because we need the project name
  if err != nil {
    return err
  }
  if err := s.populateMonitoringURL(job, project); err != nil {
    return err
  }
  return nil
}
```

Since it doesn't make sense to instantiate a service inside a model, a MLPService singleton has to be used.

The other way is to store the project name inside ensembling job which probably makes the most sense. Do you know if there are assumptions that MLP project name will always remain the same? Or can a project be renamed?

File: api/turing/batch/ensembling/runner.go
Comment (ashwinath) [reply api/turing/batch/ensembling/runner.go:125]: Yeah in hindsight this seems a little awkward. But I think this is related to https://github.com/gojek/turing/pull/87#discussion_r699061831 as that 2nd argument could just be removed completely.

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:246]: Sure.

File: api/turing/api/ensembling_job_api.go
Comment (romanwozniak) [reply api/turing/api/ensembling_job_api.go:53]: Currently project can't be renamed, but I wouldn't rely on this assumption going forward. Yeah, having a singleton service would be weird. Okay, ignore this comment then

File: api/turing/batch/ensembling/runner.go
Comment (romanwozniak) [reply api/turing/batch/ensembling/runner.go:125]: I commented on https://github.com/gojek/turing/pull/87#discussion_r699061831, would be nice to remove the second argument

Comment (ashwinath) [reply api/turing/batch/ensembling/runner.go:125]: Alright, will fetch project again inside EnsemblingJobService again.

File: api/turing/api/ensembling_job_api.go
Comment (ashwinath) [reply api/turing/api/ensembling_job_api.go:53]: Ok but I will revert this to take in projectID instead of project instead.

File: api/turing/batch/ensembling/runner.go
Comment (ashwinath) [reply api/turing/batch/ensembling/runner.go]: I have removed this

File: api/turing/batch/ensembling/runner_test.go
Comment (romanwozniak) [inline api/turing/batch/ensembling/runner_test.go:89]: Did you forget to update this mocks after reverting the signature of methods on `EnsemblingJobService`?

File: api/turing/service/ensembling_job_service.go
Comment (romanwozniak) [inline api/turing/service/ensembling_job_service.go:117]: Doesn't it make more sense to use `ensemblingJob.ProjectID` instead?

File: api/turing/service/ensembling_job_service.go
Comment (romanwozniak) [inline api/turing/service/ensembling_job_service.go:178]: Same here. Why do you need to use `options.ProjectID` instead `r.ProjectID`?

File: api/turing/service/ensembling_job_service.go
Comment (romanwozniak) [inline api/turing/service/ensembling_job_service.go:267]: `MonitoringURL` is not persisted in the database, right? Is the value remain set after `Save(job)` is called?

**Comment by romanwozniak** [comment]
> > I'm done with my review, left some comments
> 
> Ok I have addressed them @romanwozniak

Conceptually it LGTM, but there are some minor fixes that I commented about to be added before merging this.

File: api/turing/batch/ensembling/runner_test.go
Comment (ashwinath) [reply api/turing/batch/ensembling/runner_test.go:89]: hmm why didn't the test flag this out, let me check

Comment (ashwinath) [reply api/turing/batch/ensembling/runner_test.go:89]: This is not very intuitive but I think I have found out the reason why it does not give a runtime error when more arguments are supplied (not less).

Seems to have an [issue raised with the library here](https://github.com/stretchr/testify/issues/587) and it seems like it is a gotcha with the library when `mock.Anything` is used. The arity of the function is not considered when using svc.On(funcName, mock.Anything).

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:267]: It's not persisted but let's just not take the chance that the value might be wiped, so I'll put it after persisting.

File: api/turing/service/ensembling_job_service.go
Comment (ashwinath) [reply api/turing/service/ensembling_job_service.go:178]: Ok changed it.

**Comment by ashwinath** [comment]
> > > I'm done with my review, left some comments
> > 
> > 
> > Ok I have addressed them @romanwozniak
> 
> Conceptually it LGTM, but there are some minor fixes that I commented about to be added before merging this.

Updated changes.

File: api/turing/api/ensembling_job_api_it_test.go
Comment (romanwozniak) [inline api/turing/api/ensembling_job_api_it_test.go:387]: Update the mock here as well?

**Review by romanwozniak — APPROVED** [review]
LGTM, except this comment: https://github.com/gojek/turing/pull/87#discussion_r700860843. Good to go once it's fixed.

File: api/turing/api/ensembling_job_api_it_test.go
Comment (ashwinath) [reply api/turing/api/ensembling_job_api_it_test.go]: Yeah updated the rest of this file as well. Thanks for the catch.

**Pull request merged**
